solutions/hw2/task2/cacher.py

```diff
@@ -0,0 +1,83 @@
+from typing import Any, Tuple, Callable
+from collections import OrderedDict
+import functools
+
+
+def FunctionCacher(function=None, *, maximum_num_of_caches: int = 0):
+    if function:
+        return _FunctionCallsCacher(function, maximum_num_of_caches=maximum_num_of_caches)
+
+    def wrapper(func):
+        return _FunctionCallsCacher(func, maximum_num_of_caches=maximum_num_of_caches)
+
+    return wrapper
+
+
+class FunctionCallArgumentsCache:
+    nameless_args: Tuple
+    named_args: frozenset
+
+    def __init__(self, args: Tuple, kwargs: dict):
+        self.nameless_args = args
+        self.named_args = frozenset(kwargs.items())
+
+    def __hash__(self):
+        return hash((self.named_args, self.nameless_args))
+
+    def __eq__(self, other):
+        if not isinstance(other, FunctionCallArgumentsCache):
+            raise NotImplementedError("Can't compare function call argument cache to other objects")
+        return self.named_args == other.named_args and self.nameless_args == other.nameless_args
+
+
+class _FunctionCallsCacher:
+    function: Callable
+    maximum_num_of_caches: int
+    is_caching_enabled: bool
+    cache: "OrderedDict[FunctionCallArgumentsCache, Any]"
+
+    def __init__(self, function=None, *, maximum_num_of_caches):
+        self.function = function
+
+        self.maximum_num_of_caches = maximum_num_of_caches
+        self.is_caching_enabled = self.check_for_working_statement()
+
+        self.cache = OrderedDict()
+
+        functools.update_wrapper(self, function)
+
+    def check_for_working_statement(self) -> bool:
+        if self.maximum_num_of_caches < 0:
+            raise TypeError("Maximum number of caches can't be negative")
+        elif self.maximum_num_of_caches == 0:
+            return False
+        else:
+            return True
+
+    def __call__(self, *args, **kwargs):
+        result = self.function(*args, **kwargs)
```

Comment:
Should we try to find the result in the case before calculating it? If we don't, what is the point of storing the cache?